src/slice.cr

```diff
@@ -169,6 +169,48 @@ struct Slice(T)
     Slice.new(@pointer + offset, @size - offset, read_only: @read_only)
   end
 
+  # Returns a new slice that has `self`'s elements followed by *other*'s
+  # elements.
+  #
+  # ```
+  # Slice[1, 2] + Slice[3, 4, 5]          # => Slice[1, 2, 3, 4, 5]
+  # Slice[1, 2, 3] + Slice['a', 'b', 'c'] # => Slice[1, 2, 3, 'a', 'b', 'c']
+  # ```
+  #
+  # See also: `Slice.join` to join multiple slices at once without creating
+  # intermediate results.
+  def +(other : Slice) : Slice
+    Slice.join({self, other})
+  end
+
+  # Returns a new slice that has the elements from *slices* joined together.
+  #
+  # ```
+  # Slice.join([Slice[1, 2], Slice[3, 4, 5]])        # => Slice[1, 2, 3, 4, 5]
+  # Slice.join({Slice[1], Slice['a'], Slice["xyz"]}) # => Slice[1, 'a', "xyz"]
+  # ```
+  #
+  # See also: `#+(other : Slice)`.
+  def self.join(slices : Indexable(Slice)) : Slice
```

Comment:
If there is a splat overload then `Slice(Slice)` becomes ambiguous:

```crystal
def Slice.join(slices : Indexable(Slice)); end
def Slice.join(*slices : Slice); end

# first overload
Slice.join(Slice[Slice[1], Slice[2]]) # => Slice[1, 2]
Slice.join(Slice[Slice[1]])           # => Slice[1]

# second overload
Slice.join(Slice[Slice[1]], Slice[Slice[2]]) # => Slice[Slice[1], Slice[2]]
Slice.join(Slice[Slice[1]])                  # => Slice[Slice[1]]
```

Neither overload is more specific than the other, so `Slice.join(Slice[Slice[1]])`'s return value depends on the definition order.

Maybe we can require the splat overload to take 2+ arguments, i.e. `(slice : Slice, *slices : Slices)`? Note that if someone has a `Tuple` of `Slice`s already then they don't need to explode the `Tuple`.

(Actually this reasoning applies to some other methods too, such as `Object#in?`.)

Reply:
> Neither overload is more specific than the other

Is that the case?

I just tried this:

```crystal
def foo(x)
  puts "no splat"
end

def foo(*x)
  puts "yes splat"
end


foo(1)
foo(1, 2)
```

It prints "no splat" and "yes splat". Then I tried this:

```crystal
def foo(*x)
  puts "yes splat"
end

def foo(x)
  puts "no splat"
end

foo(1)
foo(1, 2)
```

Same output.

So a no-splat overload is more specific than a splat overload. I don't think there's any ambiguity, because if you do `Slice.join(...)` you would only want the splat version when there's more than one element in that list. Which... I guess means we could also make it explicit with `(slice, *slices)`.